elastic/elastic-agent
# Agent actions signature validation

## What does this PR do?

Implements the initial phase of the agent tamper protection. The scope of the change:
* Process the agent protection configuration from the policy
* Inject the agent protection configuration into the endpoint configuration
* Validate the signed agent protection configuration from the policy, and the policy id match.
* Use the signature validation key to validate the INPUT_ACTION type of actions signature only if it it present. Initial release will have only some Endpoint actions signed, but the current implementation allows all the INPUT_ACTION actions types to be signed and validated if needed.
* In addition to the action signature validation, the action id and the agent id matches are validated.
* Dispatch the action with the signature attached if it was signed.

This change doesn't include the uninstall token protection, which is scheduled for later release.

It is possible the Endpoint configuration format might change based on further collaboration with Endpoint team.

## Why is it important?

Adding signature validation for the part of the policy and the actions allows to protect the agent protection configuration and the input actions from unauthorized altering on a way from the the stack to the agent. 

## Checklist


## How to test this PR locally

1. The agent should run after upgrade as before (regression testing)
2. The unsigned action should work as before. The agent doesn't enforce signature, but only validates it if it is provided.
3. The signed action signature should be fully validated. In normal case scenario, there action signature will always be valid.

The following invalid policy or the action signature cases require manual policy or the actions manipulation:
1. The policy with invalid signature should fail to be applied, handled the same way as the bad policy configuration in the current agent.
2. The policy protected configuration that has different id than the policy should fail the policy validation
3. The agent action with invalid signature (wrong key, wrong type of the key, ...) should be acked with error pointing to the signature validation error.
4. The signed agent action data should fail the action validation as well when the signed action_id doesn't match of the action_id, when the signed agent ids do not contain the current agent id.

## Related issues

The related issue is in the private security repository, so not linking it here.

## Screenshots

Here is an example of the `"protection"` configuration added to the Endpoint configuration at the top level similar to the fleet config. This can be changed however is convenient for Endpoint to process. We could have a new top level configuration `"agent"` with `"protection"` property in it similar to the policy. @nfritts or anybody from Endpoint team.
<img width="1021" alt="Screenshot 2023-02-28 at 4 49 37 PM" src="https://user-images.githubusercontent.com/872351/223013188-1c2727ab-9900-403d-b9c1-5c1659fd4ae6.png">

Endpoint `"isolate"` signed action:
<img width="1211" alt="Screenshot 2023-03-05 at 6 59 19 PM" src="https://user-images.githubusercontent.com/872351/223013296-8dc70305-8154-46a2-88e3-6b944c44739c.png">

Endpoint isolate action response:
<img width="633" alt="Screenshot 2023-03-05 at 6 59 01 PM" src="https://user-images.githubusercontent.com/872351/223013350-11050369-417f-40e9-8fab-c0eae6e4b0c4.png">


Example of osquery action result signed with with non-matching signature (requires manual "tampering with the .fleet-actions document).

<img width="594" alt="Screenshot 2023-03-02 at 9 48 09 PM" src="https://user-images.githubusercontent.com/872351/223014100-9df01e58-4502-4914-acba-294b47bc967b.png">


## Logs

Log for invalid signature contains more details:
<img width="2267" alt="Screenshot 2023-03-02 at 9 48 35 PM" src="https://user-images.githubusercontent.com/872351/223014155-5b1527d5-bda3-4acc-bbbd-e95892918d2f.png">
